// gaze/lib/gaze.js
/*
 * gaze
 * https://github.com/shama/gaze
 *
 * Copyright (c) 2013 Kyle Robinson Young
 * Licensed under the MIT license.
 */

'use strict';

// libs
var util = require('util');
var EE = require('events').EventEmitter;
var fs = require('fs');
var path = require('path');
var globule = require('globule');
var helper = require('./helper');

// shim setImmediate for node v0.8
var setImmediate = require('timers').setImmediate;
if (typeof setImmediate !== 'function') {
  setImmediate = process.nextTick;
}

// globals
var delay = 10;

// `Gaze` EventEmitter object to return in the callback
function Gaze(patterns, opts, done) {
  var self = this;
  EE.call(self);

  // If second arg is the callback
  if (typeof opts === 'function') {
    done = opts;
    opts = {};
  }

  // Default options
  opts = opts || {};
  opts.mark = true;
  opts.interval = opts.interval || 100;
  opts.debounceDelay = opts.debounceDelay || 500;
  opts.cwd = opts.cwd || process.cwd();
  this.options = opts;

  // Default done callback
  done = done || function() {};

  // Remember our watched dir:files
  this._watched = Object.create(null);

  // Store watchers
  this._watchers = Object.create(null);

  // Store watchFile listeners
  this._pollers = Object.create(null);

  // Store patterns
  this._patterns = [];

  // Cached events for debouncing
  this._cached = Object.create(null);

  // Set maxListeners
  if (this.options.maxListeners) {
    this.setMaxListeners(this.options.maxListeners);
    Gaze.super_.prototype.setMaxListeners(this.options.maxListeners);
    delete this.options.maxListeners;
  }

  // Initialize the watch on files
  if (patterns) {
    this.add(patterns, done);
  }

  // keep the process alive
  this._keepalive = setInterval(function() {}, 200);

  return this;
}
util.inherits(Gaze, EE);

// Main entry point. Start watching and call done when setup
module.exports = function gaze(patterns, opts, done) {
  return new Gaze(patterns, opts, done);
};
module.exports.Gaze = Gaze;

// Override the emit function to emit `all` events
// and debounce on duplicate events per file
Gaze.prototype.emit = function() {
  var self = this;
  var args = arguments;

  var e = args[0];
  var filepath = args[1];
  var timeoutId;

  // If not added/deleted/changed/renamed then just emit the event
  if (e.slice(-2) !== 'ed') {
    Gaze.super_.prototype.emit.apply(self, args);
    return this;
  }

  // Detect rename event, if added and previous deleted is in the cache
  if (e === 'added') {
    Object.keys(this._cached).forEach(function(oldFile) {
      if (self._cached[oldFile].indexOf('deleted') !== -1) {
        args[0] = e = 'renamed';
        [].push.call(args, oldFile);
        delete self._cached[oldFile];
        return false;
      }
    });
  }

  // If cached doesnt exist, create a delay before running the next
  // then emit the event
  var cache = this._cached[filepath] || [];
  if (cache.indexOf(e) === -1) {
    helper.objectPush(self._cached, filepath, e);
    clearTimeout(timeoutId);
    timeoutId = setTimeout(function() {
      delete self._cached[filepath];
    }, this.options.debounceDelay);
    // Emit the event and `all` event
    Gaze.super_.prototype.emit.apply(self, args);
    Gaze.super_.prototype.emit.apply(self, ['all', e].concat([].slice.call(args, 1)));
  }

  // Detect if new folder added to trigger for matching files within folder
  if (e === 'added') {
    if (helper.isDir(filepath)) {
      fs.readdirSync(filepath).map(function(file) {
        return path.join(filepath, file);
      }).filter(function(file) {
        return globule.isMatch(self._patterns, file, self.options);
      }).forEach(function(file) {
        self.emit('added', file);
      });
    }
  }

  return this;
};

// Close watchers
Gaze.prototype.close = function(_reset) {
  var self = this;
  _reset = _reset === false ? false : true;
  Object.keys(self._watchers).forEach(function(file) {
    self._watchers[file].close();
  });
  self._watchers = Object.create(null);
  Object.keys(this._watched).forEach(function(dir) {
    self._unpollDir(dir);
  });
  if (_reset) {
    self._watched = Object.create(null);
    setTimeout(function() {
      self.emit('end');
      self.removeAllListeners();
      clearInterval(self._keepalive);
    }, delay + 100);
  }
  return self;
};

// Add file patterns to be watched
Gaze.prototype.add = function(files, done) {
  if (typeof files === 'string') { files = [files]; }
  this._patterns = helper.unique.apply(null, [this._patterns, files]);
  files = globule.find(this._patterns, this.options);
  this._addToWatched(files);
  this.close(false);
  this._initWatched(done);
};

// Dont increment patterns and dont call done if nothing added
Gaze.prototype._internalAdd = function(file, done) {
  var files = [];
  if (helper.isDir(file)) {
    files = [helper.markDir(file)].concat(globule.find(this._patterns, this.options));
  } else {
    if (globule.isMatch(this._patterns, file, this.options)) {
      files = [file];
    }
  }
  if (files.length > 0) {
    this._addToWatched(files);
    this.close(false);
    this._initWatched(done);
  }
};

// Remove file/dir from `watched`
Gaze.prototype.remove = function(file) {
  var self = this;
  if (this._watched[file]) {
    // is dir, remove all files
    this._unpollDir(file);
    delete this._watched[file];
  } else {
    // is a file, find and remove
    Object.keys(this._watched).forEach(function(dir) {
      var index = self._watched[dir].indexOf(file);
      if (index !== -1) {
        self._unpollFile(file);
        self._watched[dir].splice(index, 1);
        return false;
      }
    });
  }
  if (this._watchers[file]) {
    this._watchers[file].close();
  }
  return this;
};

// Return watched files
Gaze.prototype.watched = function() {
  return this._watched;
};

// Returns `watched` files with relative paths to process.cwd()
Gaze.prototype.relative = function(dir, unixify) {
  var self = this;
  var relative = Object.create(null);
  var relDir, relFile, unixRelDir;
  var cwd = this.options.cwd || process.cwd();
  if (dir === '') { dir = '.'; }
  dir = helper.markDir(dir);
  unixify = unixify || false;
  Object.keys(this._watched).forEach(function(dir) {
    relDir = path.relative(cwd, dir) + path.sep;
    if (relDir === path.sep) { relDir = '.'; }
    unixRelDir = unixify ? helper.unixifyPathSep(relDir) : relDir;
    relative[unixRelDir] = self._watched[dir].map(function(file) {
      relFile = path.relative(path.join(cwd, relDir) || '', file || '');
      if (helper.isDir(file)) {
        relFile = helper.markDir(relFile);
      }
      if (unixify) {
        relFile = helper.unixifyPathSep(relFile);
      }
      return relFile;
    });
  });
  if (dir && unixify) {
    dir = helper.unixifyPathSep(dir);
  }
  return dir ? relative[dir] || [] : relative;
};

// Adds files and dirs to watched
Gaze.prototype._addToWatched = function(files) {
  for (var i = 0; i < files.length; i++) {
    var file = files[i];
    var filepath = path.resolve(this.options.cwd, file);

    var dirname = (helper.isDir(file)) ? filepath : path.dirname(filepath);
    dirname = helper.markDir(dirname);

    // If a new dir is added
    if (helper.isDir(file) && !(filepath in this._watched)) {
      helper.objectPush(this._watched, filepath, []);
    }

    if (file.slice(-1) === '/') { filepath += path.sep; }
    helper.objectPush(this._watched, path.dirname(filepath) + path.sep, filepath);

    // add folders into the mix
    var readdir = fs.readdirSync(dirname);
    for (var j = 0; j < readdir.length; j++) {
      var dirfile = path.join(dirname, readdir[j]);
      if (fs.statSync(dirfile).isDirectory()) {
        helper.objectPush(this._watched, dirname, dirfile + path.sep);
      }
    }
  }
  return this;
};

Gaze.prototype._watchDir = function(dir, done) {
  var self = this;
  var timeoutId;
  try {
    this._watchers[dir] = fs.watch(dir, function(event) {
      // race condition. Let's give the fs a little time to settle down. so we
      // don't fire events on non existent files.
      clearTimeout(timeoutId);
      timeoutId = setTimeout(function() {
        // race condition. Ensure that this directory is still being watched
        // before continuing.
        if ((dir in self._watchers) && fs.existsSync(dir)) {
          done(null, dir);
        }
      }, delay + 100);
    });
  } catch (err) {
    return this._handleError(err);
  }
  return this;
};

Gaze.prototype._unpollFile = function(file) {
  if (this._pollers[file]) {
    fs.unwatchFile(file, this._pollers[file] );
    delete this._pollers[file];
  }
  return this;
};

Gaze.prototype._unpollDir = function(dir) {
  this._unpollFile(dir);
  for (var i = 0; i < this._watched[dir].length; i++) {
    this._unpollFile(this._watched[dir][i]);
  }
};

Gaze.prototype._pollFile = function(file, done) {
  var opts = { persistent: true, interval: this.options.interval };
  if (!this._pollers[file]) {
    this._pollers[file] = function(curr, prev) {
      done(null, file);
    };
    try {
      fs.watchFile(file, opts, this._pollers[file]);
    } catch (err) {
      return this._handleError(err);
    }
  }
  return this;
};

// Initialize the actual watch on `watched` files
Gaze.prototype._initWatched = function(done) {
  var self = this;
  var cwd = this.options.cwd || process.cwd();
  var curWatched = Object.keys(self._watched);

  // if no matching files
  if (curWatched.length < 1) {
    // Defer to emitting to give a chance to attach event handlers.
    setImmediate(function () {
      self.emit('ready', self);
      if (done) { done.call(self, null, self); }
      self.emit('nomatch');
    });
    return;
  }

  helper.forEachSeries(curWatched, function(dir, next) {
    dir = dir || '';
    var files = self._watched[dir];
    // Triggered when a watched dir has an event
    self._watchDir(dir, function(event, dirpath) {
      var relDir = cwd === dir ? '.' : path.relative(cwd, dir);
      relDir = relDir || '';

      fs.readdir(dirpath, function(err, current) {
        if (err) { return self.emit('error', err); }
        if (!current) { return; }

        try {
          // append path.sep to directories so they match previous.
          current = current.map(function(curPath) {
            if (fs.existsSync(path.join(dir, curPath)) && fs.statSync(path.join(dir, curPath)).isDirectory()) {
              return curPath + path.sep;
            } else {
              return curPath;
            }
          });
        } catch (err) {
          // race condition-- sometimes the file no longer exists
        }

        // Get watched files for this dir
        var previous = self.relative(relDir);

        // If file was deleted
        previous.filter(function(file) {
          return current.indexOf(file) < 0;
        }).forEach(function(file) {
          if (!helper.isDir(file)) {
            var filepath = path.join(dir, file);
            self.remove(filepath);
            self.emit('deleted', filepath);
          }
        });

        // If file was added
        current.filter(function(file) {
          return previous.indexOf(file) < 0;
        }).forEach(function(file) {
          // Is it a matching pattern?
          var relFile = path.join(relDir, file);
          // Add to watch then emit event
          self._internalAdd(relFile, function() {
            self.emit('added', path.join(dir, file));
          });
        });

      });
    });

    // Watch for change/rename events on files
    files.forEach(function(file) {
      if (helper.isDir(file)) { return; }
      self._pollFile(file, function(err, filepath) {
        // Only emit changed if the file still exists
        // Prevents changed/deleted duplicate events
        if (fs.existsSync(filepath)) {
          self.emit('changed', filepath);
        }
      });
    });

    next();
  }, function() {

    // Return this instance of Gaze
    // delay before ready solves a lot of issues
    setTimeout(function() {
      self.emit('ready', self);
      if (done) { done.call(self, null, self); }
    }, delay + 100);

  });
};

// If an error, handle it here
Gaze.prototype._handleError = function(err) {
  if (err.code === 'EMFILE') {
    return this.emit('error', new Error('EMFILE: Too many opened files.'));
  }
  return this.emit('error', err);
};
²è!r¤–l¸âY—›bƒt“ëh¾I2ü&ÙÜÅ^S<$¯cœ #|£§ZşH<j:¢½ÔJOˆ|™ä=(+ÎšsuGB;µ§zr&‚)b¼]¹º°™­¤[’I[/8Ñù×İZ¹£ĞÊ úrDHı÷pÍú|@!ôA1é¹ ”®Ùå¥W ;@:á’öäï~²?#$%‘„*µã£İ/BI³®ßçq‹_/	ëõ“w=æıIƒ	dX{#ÁÍâ@IvÔŞPòş¦ŞO|7ìˆÊ¢ë¯É3hi)\'¬¬)µ½k‰¤?;Mfµs¼‹>ğ¥/!<¬Ó ükîÿKu»H'üc!ŞÙ\cî9×¥9¤ìË"V½©³JÏ,C,Pñºs]°€ëú­ö\ë…ô-ñl•ï×§6RœLº>ş‰–">f~ÜLñ¿hK2t ™ÏóI°”îôüAº¯Êøä6;ˆ 9ZØ÷É.GÃfYA‡OÃ©Ğ4ÉœíÁÉ
"ÛŞİ‰¿Ï’Sn€ß2Ç %Z„2åd³™zÆ6…cPÓu:â=$ ™4/Ç†$wğ =4BÎõ¹éI²ƒÕfÒ&R‹Á»¦†vÖD7kò.¸|õHÃJ-ÖÈz¦ù$Q¹ÖËc5g[ïü”jmûF¦étµ•üõ^Æ$büT³çœmÇSª9¬w”iÌ=rÖé÷RÓo\J6JşvÛáäUc4œûÛ™Ëò¬ˆ¯ˆËjrë}@£æ[°ÑL]®ùÕš‰¡×&Í¸ 
õb;Ø[	?¼J’NëOMğ·}`T›ÌL—?àÌ3Õ™ª3¥y•L¶>‚b¸RN<]Xí½~ÁgBƒèÌ\	ıVì~6 èoª¤šÔ¤úì>2Vhİvø4öğë4&}À( ÷©Ê?´ÎŠ_8¥L¨£òí÷Ş€«­oR=1†´
mm¹Æ6`Cñ‡óK-ªñš%¹Ÿú^d½‘a^³¥ìdá¤:åÒ¶RJİéw”kp>ÍD@
?PÄqvãĞš›dÔ®{Óh¶vl¤S,eÕ6`t˜·(„ÇÌõÊ¤e Y~/î–JR©ª2â(–ß£Ÿº¶ë=¾¾Th¬`)æ”£F°M	Ñ[ÌMïÅìÛy¦i²Ïdáæm÷9XãáJÄƒò–<÷ “Sa#,¢ôÌY&Üd7x}p7Ä­6~k¤£‚İ–+ªà¿
«\‰Ãá $gŒ×¯wC|ƒAç°`XÊd°yxT–•	F4cpm;”·[4lc
ÑâÎ˜•+ô‚†
pÿ-SğGÒYÎ´­Ï¦ş²º;·dncNoWÉ»M¬´­DÂ²õüXÖÊQòh´qå@}äwáÛ¯Áp
hÕÈÀ°Ôlä¿~b–}â{´‚rÆÚÉÿº‹<ø÷B§•‰Ä|¥mÃ#†
g><vDXJ‚Ë¡ş«!woÇßšíh¢Õ*ÕÑ†£Šû˜S÷wU»¦ÒãCª*£¡‰¿ÊªĞ4`j¿ n`bJè$b<Ãô·Y kVµ²üŞ=yÖ 6‰£:0dĞ
4Ä—ëº Æ¼6ÖcYJ9ÍAögÎ¡tHëBävŠxoK¡ë
µfû†mšïñtKx(”ˆéò½Ó 9r‹j’2¾"”ÉœÇM^áä¾4;;†eG Ş›¯ış,8Té^¡ŸÉODšıs[~î×}Jù›6–ñæ½¼ê:¿¤ö»ƒn*cK7,ëÚ”æÅH5
’¯¨÷nEª9xA»Ø›+åÙúÿ(èÊØ‘ºˆÂmCÎ§…HÿCW%2eª­ ¯dF»s!ğ‰÷¬Ô©E_ö†Húåbƒ·)õg$H
VíÅŠ¯Ü×¬¯‘9 ê[ˆÅ	yÛ^÷«é^—Ñt3/¢ İ5ŒHqVÃ:EÅÂâ„!+ò&²Çí“¿­Æ#úcÕw+*~KA
«ßeƒÏ&–GÌ®î‚gæs"ZËz©¤,yB1­-õY^¼swp«èŸB®Èìd¿¼à'Qø”Ã‚èËÖFÍlãœ¹<¶şKmë:¥ÄéÛläUxDHu¸,Qè‹;ŒŠ«ƒèô$®;Jh*UÌŞÇJH³µy?¶	ò{›&Ó£ÇÑæôŠ#,[.Ç„g5¬İ®„ØÃr[ |jLqQŸŞß/‘yAu`1Úõú¿Ô—ka	;v–Oœ"ã·‹i»ŠE“J[„<¼Æ6H¤.ì©çF‚›ZeHíLMñA;ß8tÎn%´ıİÚ~1ğp‰°ıô|«[¶ƒŞOíF­ÌĞÂÖ.ohÿ}±¢¦¶·–°B5.c·Ï­90(Ù„r,î´¤ˆÚ"1¥äNahGï´[(èí¹ˆ¯’è°Îî^R‘@û4D'ùáÔA‘7€gg3¦Ç°àb¡²Íë#0ÅYW„“êÍß€axk.±Şµ( ïÂätYØêbF
è«èÅÒÎT‹Ä97üHó
Jã(Õ;W4	€£Ô#¸k^Z¸ ]*n7p	Î}7ãô¿VHX¬’¯OŒ­İÖ(^|Kpí¹8Œúë;Y=ùîI8ŸøŸÚºk(<Ib}™
¨Sã.WÁ(BËŞš$ÁĞ:;¡%»RÛU¶¥c_¼^¯å)–á6©ñğûP^–ÒÁãcç?M§I!YCMà¼5ÂÌPHœW³«ªD­c¡<A¥´q"|_â¬±ºƒ{#Lõë³ÊH\õÿÅw,ÖE¾¹Õ 9RÉ;ìÇIëÒ¢-eÈ#µÉÆ
Wotü¨¹•óKB5ÿbÉT–³ÙzÂpH¹X›PÛÙôIUÁ¶¼ùø¼q‡Õ´À!±å(¶ÿ$¡ÇÇæ¨aPHtùÔ{dWœ‘$Kp¢×ä'£€9üß@mpöqò{Õa[dešÎİòƒO×¸zÂ&Æ	l²zbRá~ \iT,<O­ĞBŠIwZÏŠušš*TŞgú­²Ki{‹ó…ú'©Äôs×h¿ÿíŞp^ÊaéÏr¤±p$ü #w~‹Ÿµp	‚–€ÜtO!w¼ö¦ =ì²!×ìÜ—× _³ÊDìA–‡jï¾f^÷Ez%§Ï]áDşÄ®By·5I¯ÊÑ[·ƒZº7©"„"·/>g3

äĞd¿æ¶oY%xú>èö_rúµå¼s›aA>ƒŞ´µ{§ı®	^^Ø«,=ı‹U­-7ª‘3ÿr°ör¼êıìiÕĞÉû‘§9Ò¶v…L ­KŠüã¶ÁıË=«#­xØã³À+ó´i=šJä¾Q˜,·›ıXE,Ìk{nî’ÔXúæª<"ÆB+fH}
b¦Ì,4)‡dí:E›£í•B69n‘¾:Ñ[©Æ;•EE­ĞPM„ÈoğNÓŒ¥­h³ˆYó>He³ÿ|‘Xa.äö‡%u}:T™İjêÿ¾#ç¢4We´j;ò	ÔõtÒT¼âuÚúÎ™¸¯İ¡¼NÎSl<¡Ó:¼§2h½çÈ§°{/W-wš|ÿMw›‡¢é3%_îOÿ°P—µ
Y‡¥Î`-Ë¡Ú0Èó¼úìx>ÁpB>FÉ.ê–_«ÇŸ¢‰œ;¯øº‘moY½³xí|.g5«ëSÊd¡¶Şù˜‘fÄ5,åˆ†x^t„ÍÎ'Ñ#e!˜C]mõ2º‡m^"ğTn§Ãß/MA¯Vk6¶™ĞºK=x›ç1á àóÑ-6öâ|F£[Ps
ÅÓ¿Ä³tÑÛ×^£ßÖô—»f0@f€NáºSt hw“BAÓ&³¤âpLß,6m¶ì‚¹Ç3Â‰O=tÀ,ã]nŞµe3Õj,¤¥-;Ç9şŠÙÑ6¢…¥Ÿ´åY­¡Èg™²?´$7´IäÛÃCDè]hÔƒî¯ËX±{…n·	¡îümS¦öë•Ø€mî¦imRY@’…ùb½äÀøRw'|Ì¦¹Ğ¾wŒß\ÓûÍ¶Ë+ªoi%2!°gjõWàËûÎl£ÿe®»Dû |D+‰«Z‚®*"6\»OÙ>ˆ‚˜JGŞÉ£—û§}¿üpİ­èz\‹¿LR…µm3ßiá¥ïÏÊäÑôäµE½x.’×Üö¸8·-S™0'èµãcX™ öT44}øğ®øî{ÿéYN—»œx‚Ö/8ô-u‰Xûg¾n¬…UÜ§Òº~ÁbyT*É[üGl×:“KÏD¶æ²;‰„;ÜAsOŒV±ò·+ZªIù‹>Yß‰Ïi|ğ´kÂdèÀ‚$†Z•ıR? _d«£f¥zš5Å6b™Ç$ğŒh~ÑiQ[³Ö³äŸÅT_Õ'¥¹Òw´"î}1Hk‘³)næïßÍWx¸Êj®M\q P3FÿTËOpqtsÈÅ,çÆ%œ·³+¡å»Ó¹~‚"qv.ÿšpn?…&d'gt5…W¿R”Èù§`Ğ‘#µÿƒsºDÈ^CÉØÀsXâÍ¥-5ò°íu¾Öµ7  ‹	¿Á@­É±9æï%¹	,¦½‹$ÛGm)­¢~öR±/:¥š·^J’ßgã‚.·<.$êM¦ÔÇ4œtß÷5q&	’×!ÕŞÊø™7]ƒG;$)Û¼¼FÕ‹R¸KlGüú ƒ`û@grª¸Ñµ„¡§O«9ÿL`Z&ˆƒ­Ó½ÍAƒ¸›
\•ğí m¤pŸvúûßYˆÚ áé7¶÷KşA¡¤‰d\‹)¹ÛOO“µæQ829Ü´4ÿ³xnUh7zÉnùé	Ğ•®£ÄñÈıÊ$´È¨š^•~eôŞ?Û¼)
1hÆIŸ	·wÂ„ñ<¢NJOâ,´TFø
dU´g
õ¥”f$
—ûJˆ/F_Â3Ÿb4ö‚kÂ_zŠUE) YÔ`m€©ÏËAÆ’:y¥ıxFÖîbW­}[^Â«®e†l¬®tiP1-Â§S5s0İ·ôŠ»…*ÄuwAp¨ü8hLiÂ(­ŒØg9¶´y…å÷Ëƒ¶f’ôËA~}#=‡iM&-?åW-¿ÒÈk¨	yŞe<–Hï&èë99Ù}/ôÜKP‡Ôw‚&¾DL‡T'x+‡B¦Ï¯¯•îb”Œ”d-iv‹Š5¨üb ğ™GÓè”+KCY{8Ğ›êõ.IYó@ã“>ÑQ'<øÇ×%\úÈ'PÛq%ƒJß¸sáb£ñA
